src/v/coproc/wasm_event.cc

```diff
@@ -107,40 +138,53 @@ wasm::errc validate_event(const model::record& r) {
     if (auto errc = std::get_if<wasm::errc>(&action)) {
         return *errc;
     }
+
+    auto get_type_res = get_coproc_type(r);
+    if (auto err = std::get_if<wasm::errc>(&get_type_res)) {
+        return wasm::errc::unexpected_coproc_type;
+    }
+    header.type = std::get<coproc_type>(get_type_res);
+
     /// Technically all a 'remove' event needs to be valid, is a non-empty
     /// key field
     auto wasm_action = std::get<event_action>(action);
     if (wasm_action == event_action::deploy) {
         if (r.value_size() == 0) {
             return wasm::errc::empty_mandatory_field;
         }
+        header.actoin = event_action::deploy;
+
         return verify_event_checksum(r);
     } else {
         if (r.value_size() > 0) {
             /// A 'remove' should have an empty body
             return wasm::errc::unexpected_value;
         }
+
+        header.actoin = event_action::remove;
     }
     /// A description isn't neccessary for either event type
     return wasm::errc::none;
 }
 
-absl::btree_map<script_id, iobuf>
-reconcile_events(std::vector<model::record_batch> events) {
-    absl::btree_map<script_id, iobuf> wsas;
+event_batch reconcile_events(std::vector<model::record_batch> events) {
+    event_batch wsas;
     for (auto& record_batch : events) {
         record_batch.for_each_record([&wsas](model::record r) {
-            auto mb_error = wasm::validate_event(r);
+            parsed_event new_event;
+            auto mb_error = wasm::validate_event(r, new_event.header);
             if (mb_error != wasm::errc::none) {
                 vlog(
                   coproclog.error,
                   "Erranous coproc record detected, issue: {}",
                   mb_error);
                 return;
             }
-            auto id = wasm::get_event_id(r);
+            auto id = wasm::get_event_id(r).value();
+            new_event.data = r.share_value();
             /// Update or insert, preferring the newest
-            wsas.insert_or_assign(*id, r.share_value());
+            wsas[new_event.header.type].insert_or_assign(
```

Comment:
ids must be globally unique, this is a compacted topic

Reply:
done